---
title: "PanDA application to cancer cell line data"
author: "Muhammad Aminu"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_depth: 2
    toc_float: yes
package: PANDA
vignette: |
  %\VignetteIndexEntry{Pan-omics Discriminant Analysis with PanDA}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = FALSE, 
  warning = FALSE,
  message = FALSE
)
```

```{r setup, message=FALSE, warning=FALSE}
library(Seurat)
library(Signac)
library(ggplot2)
library(cowplot)
library(caret)
library(ggpubr)
library(corrplot)
library(EnsDb.Hsapiens.v79)
library(MASS)
library(Rtsne)
library(umap)
library(pheatmap)
library(RGCCA)
library(ConsensusClusterPlus)
library(mclust)
library(aricode)
library(PANDA)
library(RColorBrewer)
```


# Introduction

In this vignette we will demonstrate how to run multi-omics discriminant analysis on single-cell datasets using the PanDA framework

# Load data

For this demo we will use a synthetic dataset simulating a scRNA-seq and scATAC-seq experimentts, generated using [MOSim](https://https://github.com/ConesaLab/MOSim).

```{r}
data("data_CancerCellLine")

rna <- data_CancerCellLine[["rna"]]
atac <- data_CancerCellLine[["atac"]]
metaData <- data_CancerCellLine[["metadata"]]
```

# Create SingleCellExperiment object

We can first initialize a Seurat object with the single-cell RNA-seq data and then apply the PanDA integration method to harmonize and enhance the analysis, enabling more robust and comprehensive insights into the dataset.

```{r, message=FALSE, warning=FALSE}
cellLine.rna <- CreateSeuratObject(counts = rna,
                                project = "cellLine",
                                assay = "RNA",
                                meta.data = metaData)

cellLine.atac <- CreateSeuratObject(counts = atac,
                                 project = "cellLine",
                                 assay = "ATAC",
                                 meta.data = metaData)
```
# Extract discriminant components

We next extracted discriminant latent components using PANDA. We assessed PanDA in terms of discriminant latent component recovery/estimation, in other words, how well it determine a common latent space that best captures the discriminant structure in the data.

```{r, message=FALSE, warning=FALSE}
data <- list(rna = as.matrix(rna), atac = as.matrix(atac))
gnd <- as.numeric(as.factor(cellLine.rna@meta.data[["orig.ident"]]))
numComponents <- 100

PanDAModel <- PanDA(data,gnd,numComponents,0.005)

PanDARNA <- CreateDimReducObject(
  embeddings = PanDAModel[["PanDAComponents"]][["rnaComponents"]],
  loadings = PanDAModel[["projMatrices"]][["Wrna"]],
  stdev = numeric(),
  key = "DC_",
  assay = "RNA"
)

cellLine.rna@reductions[["PanDA"]] <- PanDARNA

PanDAATAC <- CreateDimReducObject(
  embeddings = PanDAModel[["PanDAComponents"]][["atacComponents"]],
  loadings = PanDAModel[["projMatrices"]][["Watac"]],
  stdev = numeric(),
  key = "DC_",
  assay = "ATAC"
)

cellLine.atac@reductions[["PanDA"]] <- PanDAATAC
```
# Investigate PanDAs discriminant components
PANDA-extracted latent components can also be used as inputs to nonlinear manifold learning approaches such as t-distributed stochastic neighbor embedding and UMAP to inspect variations captured by each component as well as for other downstream analyses. We embedded cells (defined based on the 10 PANDA latent components) into lower dimensional UMAP subspaces.

```{r}
cellLine.rna <- RunUMAP(cellLine.rna, dims = 1:100, reduction = "PanDA", reduction.name = "PanDAumap", reduction.key = "PandaUMAP_")
cellLine.atac <- RunUMAP(cellLine.atac, dims = 1:100, reduction = "PanDA", reduction.name = "PanDAumap", reduction.key = "PandaUMAP_")

cols <- c("#E58601","#4E9F50","#027B8E")
p1 <- DimPlot(cellLine.rna, reduction = "PanDAumap", group.by = "orig.ident", label = FALSE, pt.size = 2, cols = cols) + NoLegend() + ggtitle("RNA")
p2 <- DimPlot(cellLine.atac, reduction = "PanDAumap", group.by = "orig.ident", label = FALSE, pt.size = 2, cols = cols) + ggtitle("ATAC")
p1 / p2
```

We then examined the components variabilities by visualizing the contribution of each component in the HeLa, K562, and HCT cancer cell lines. Both components 1 and 2 explained variation in the three cell lines, capturing differences in these groups. For example, component 1 (for both scRNA-seq and scATAC-seq modalities) assigned negative values, positive values, and values close to zero to the HCT, K562, and HeLa cells, respectively

```{r}
FeaturePlot(object = cellLine.rna, reduction = "PanDAumap", features = c('DC_1','DC_2'), pt.size = 2)&
  scale_colour_gradientn(colours = rev(brewer.pal(n = 11, name = "RdBu"))) # consider DC_1, DC_2, DC_5, DC_8
```

```{r}
FeaturePlot(object = cellLine.atac, reduction = "PanDAumap", features = c('DC_1','DC_2'), pt.size = 2)&
  scale_colour_gradientn(colours = rev(brewer.pal(n = 11, name = "RdBu")))
```

Examining the extracted PANDA latent components further, we found that these omics-specific components were not heavily correlated, indicating they captured independent information

```{r}
# Check correlation among extracted components
refData <- PanDAModel[["PanDAComponents"]][["rnaComponents"]]
M = cor(refData[,c(1:10)])

# corrplot(M)
corrplot.mixed(M)
```

we identified genes and peaks with the greatest enrichment (absolute weight) in individual latent components by examining the PANDA projection matrix. The top 30 genes with the largest absolute weights on component 1 captured differences in expression signatures that distinguished the three cancer cell lines. PANDA identified and assigned larger weights (greater importance) to the most discriminating genes in this dataset.

```{r}
# Plot gene loading for PanDA components (RNA)
refComponents <- c("DC_1","DC_2")
nfea <- 30
loadingsRNA <- as.data.frame(cellLine.rna@reductions[["PanDA"]]@feature.loadings[,refComponents])
rownames(loadingsRNA) <- rownames(rna)
tempfeaRNA <- data.frame(matrix(ncol = length(refComponents), nrow = nfea))
colnames(tempfeaRNA) <- refComponents

genes <- data.frame(matrix(ncol = length(refComponents), nrow = nfea))
colnames(genes) <- refComponents

signRNA <- data.frame(matrix(ncol = length(refComponents), nrow = nfea))
colnames(signRNA) <- refComponents

for (i in 1:length(refComponents)) {
  ind <- sort(abs(loadingsRNA[,i]), decreasing = TRUE, index.return=TRUE)$ix
  tempfeaRNA[,i] <- loadingsRNA[ind[c(1:nfea)],i]
  genes[,i] <- rownames(loadingsRNA)[ind[c(1:nfea)]]
  for (j in 1:nrow(tempfeaRNA)){
    if(tempfeaRNA[j,i] > 0){
      signRNA[j,i] <- "+"
    }
    else{
      signRNA[j,i] <- "-"
    }
  }
}

# Plot the genes with top weights on component 1
ffeaRNA <- data.frame(matrix(ncol = 3, nrow = nfea))
component <- "DC_1"
ffeaRNA[,1] <- genes[,component]
ffeaRNA[,2] <- abs(tempfeaRNA[,component])
ffeaRNA[,3] <- signRNA[,component]
colnames(ffeaRNA) <- c("genes","weights","sign")

geneIDS <- ensembldb::select(EnsDb.Hsapiens.v79, keys= ffeaRNA[,1], keytype = "GENEID", columns = c("SYMBOL","GENEID"))
ffeaRNA$renamedGenes <- geneIDS$SYMBOL

ggdotchart(ffeaRNA, x = "renamedGenes", y = "weights",
           palette = cols,
           sorting = "descending",
           add = "segments",
           add.params = list(color = "#999999", size = 3),
           rotate = TRUE,
           dot.size = 6,
           label = ffeaRNA$sign,
           font.label = list(face = "bold", color = "white", size = 10, 
                             vjust = 0.4), 
           ggtheme = theme_pubr(),
           xlab = "Genes"
)+
  font("x.text", size = 11, color = "black", face = "bold.italic")+
  font("y.text", size = 11, color = "black", face = "bold.italic")+
  font("xy", size = 11, color = "black", face = "bold.italic")

```

Heat map of the clustering analysis using these 30 genes on the first PANDA latent components optimally highlighted clusters corresponding to the three cancer cell lines while depicting the sparsity commonly seen in scRNA-seq and scATAC-seq datasets

```{r}
# plot heatmap with the top genes on component 1
nredRNA <- rna[genes[,component],]
namesredRNA <- rownames(nredRNA)

geneID2 <- ensembldb::select(EnsDb.Hsapiens.v79, keys= namesredRNA, keytype = "GENEID", columns = c("SYMBOL","GENEID"))
rownames(nredRNA) <- geneID2$SYMBOL

# out <- pheatmap(scale(nredRNA), cluster_rows=T, show_rownames=F, cutree_rows = 3, cutree_cols = 2)
outRNA <- pheatmap(t(nredRNA), cluster_rows=T, scale="row", show_rownames=F, cutree_rows = 3, cutree_cols = 2)

# Add row and column annotations
rnalabel <- cutree(outRNA$tree_row, k=3)[outRNA$tree_row[["order"]]]
annot_rna <- data.frame(row.names = names(rnalabel),
                        Clusters = as.factor(rnalabel),
                        Groundtruth = as.factor(metaData[outRNA[["tree_row"]][["order"]],"celltype"]))

my_colour = list(
  Groundtruth = c(HCT = "#E58601", Hela = "#4E9F50", K562 = "#027B8E"),
  clustering = c("1" = "#CD0BBC", "2" = "#F5C710", "3" = "#28E2E5"))

pheatmap(t(nredRNA), cluster_rows = T, cluster_cols = T,
         show_rownames = F, main = "RNA",
         scale = "row", show_colnames = T, 
         cutree_rows = 3, annotation_row = annot_rna,
         annotation_colors = my_colour)

```

These 30 genes collectively yielded ARI and NMI values of 1.00

```{r}
rnaclustres <- clustComp(annot_rna$Cluster, factor(annot_rna$Groundtruth))
cat(" ARI =", rnaclustres[["ARI"]],"\n", "NMI =", rnaclustres[["NMI"]])
```



<details>
  <summary>**Session Info**</summary>
  
```{r}
sessionInfo()
```

</details>
